collect_app/src/main/java/org/odk/collect/android/external/FormUriActivity.kt

```diff
@@ -51,130 +65,41 @@ class FormUriActivity : ComponentActivity() {
+    private val formUriViewModel by viewModels<FormUriViewModel> {
+        object : ViewModelProvider.Factory {
+            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
+                return FormUriViewModel(
+                    intent.data,
+                    scheduler,
+                    projectsRepository,
+                    projectsDataService,
+                    contentResolver,
+                    formsRepositoryProvider,
+                    instanceRepositoryProvider,
+                    resources
+                ) as T
             }
-        } ?: false
-
-        return if (!isUriValid) {
-            displayErrorDialog(getString(org.odk.collect.strings.R.string.unrecognized_uri))
-            false
-        } else {
-            true
         }
     }
 
-    private fun assertFormExists(): Boolean {
-        val uri = intent.data!!
-        val uriMimeType = contentResolver.getType(uri)
-
-        val doesFormExist = if (uriMimeType == FormsContract.CONTENT_ITEM_TYPE) {
-            formsRepositoryProvider.get().get(ContentUriHelper.getIdFromUri(uri))?.let {
-                File(it.formFilePath).exists()
-            } ?: false
-        } else {
-            instanceRepositoryProvider.get().get(ContentUriHelper.getIdFromUri(uri))?.let {
-                if (!File(it.instanceFilePath).exists()) {
-                    Analytics.log(AnalyticsEvents.OPEN_DELETED_INSTANCE)
-                    InstanceDeleter(instanceRepositoryProvider.get(), formsRepositoryProvider.get()).delete(it.dbId)
-                    displayErrorDialog(getString(org.odk.collect.strings.R.string.instance_deleted_message))
-                    return false
-                }
-
-                val candidateForms = formsRepositoryProvider.get().getAllByFormIdAndVersion(it.formId, it.formVersion)
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        DaggerUtils.getComponent(this).inject(this)
+        setContentView(R.layout.circular_progress_indicator)
 
-                if (candidateForms.isEmpty()) {
-                    val version = if (it.formVersion == null) {
-                        ""
-                    } else {
-                        "\n${getString(org.odk.collect.strings.R.string.version)} ${it.formVersion}"
+        formUriViewModel.error.observe(this) {
+            if (it == null) {
+                if (savedInstanceState != null) {
+                    if (!savedInstanceState.getBoolean(FORM_FILLING_ALREADY_STARTED)) {
+                        startForm()
                     }
-
-                    displayErrorDialog(getString(org.odk.collect.strings.R.string.parent_form_not_present, "${it.formId}$version"))
-                    return false
-                } else if (candidateForms.count { form: Form -> !form.isDeleted } > 1) {
```

Comment:
I've noticed that you removed this counting only not deleted forms. Is that not needed anymore?

Reply:
It could be that there's a test missing. Do you have a sense of the case that code was for?

Reply:
Yes, I think it was for the case where you try to open an instance that is marked as deleted in the database. It shouldn't be doable from ODK Collect but if you use an external app and content providers then it might be possible.

Reply:
We actually do let you open an instance for a form that's been deleted - there's a test for it at `DeleteBlankFormTest#deletingAForm_whenThereFilledForms_removesFormFromBlankFormList_butAllowsEditingFilledForms`.

I realised that the code here is actually checking if an instance has multiple form definitions **that aren't deleted**. I'll fix this.